File: client/src/services/api/foodEntries.ts
import { apiFetch } from "./http";
import type { Unit } from "./portions";
import { parseNumeric } from "@services/utils/parsing";

export type MealType = "breakfast" | "lunch" | "dinner" | "snacks" | "water";

// API response type (Decimal amount comes as string from backend)
type FoodEntryApiResponse = {
  id: string;
  product_id: string;
  portion_id: string;
  day: string; // YYYY-MM-DD
  meal_type: MealType;
  amount: string | number; // Decimal from backend
  unit: Unit;
  created_at: string;
  updated_at: string;
};

// Frontend type with parsed number
export type FoodEntry = {
  id: string;
  productId: string;
  portionId: string;
  day: string; // YYYY-MM-DD
  mealType: MealType;
  amount: number;
  unit: Unit;
  createdAt: string;
  updatedAt: string;
};

export type FoodEntryCreateRequest = {
  product_id: string;
  portion_id: string;
  day: string; // YYYY-MM-DD
  meal_type: MealType;
  amount: string | number;
  unit: Unit;
};

export type FoodEntryUpdateRequest = Partial<{
  portion_id: string;
  meal_type: MealType;
  amount: string | number;
  unit: Unit;
}>;

// Transform API response to frontend type
const transformFoodEntryResponse = (
  response: FoodEntryApiResponse,
): FoodEntry => {
  const amount = parseNumeric(response.amount);
  if (amount === undefined) {
    throw new Error(`Invalid amount value in food entry: ${response.amount}`);
  }

  return {
    id: response.id,
    productId: response.product_id,
    portionId: response.portion_id,
    day: response.day,
    mealType: response.meal_type,
    amount,
    unit: response.unit,
    createdAt: response.created_at,
    updatedAt: response.updated_at,
  };
};

export const listFoodEntries = async (query?: {
  day?: string;
  from?: string;
  to?: string;
}): Promise<FoodEntry[]> => {
  const responses = await apiFetch<FoodEntryApiResponse[]>("/v1/food-entries", {
    query,
  });
  return responses.map(transformFoodEntryResponse);
};

export const createFoodEntry = async (
  body: FoodEntryCreateRequest,
): Promise<FoodEntry> => {
  const response = await apiFetch<FoodEntryApiResponse>("/v1/food-entries", {
    method: "POST",
    body,
  });
  return transformFoodEntryResponse(response);
};

export const getFoodEntry = async (entryId: string): Promise<FoodEntry> => {
  const response = await apiFetch<FoodEntryApiResponse>(
    `/v1/food-entries/${entryId}`,
  );
  return transformFoodEntryResponse(response);
};

export const updateFoodEntry = async (
  entryId: string,
  patch: FoodEntryUpdateRequest,
): Promise<FoodEntry> => {
  const response = await apiFetch<FoodEntryApiResponse>(
    `/v1/food-entries/${entryId}`,
    {
      method: "PATCH",
      body: patch,
    },
  );
  return transformFoodEntryResponse(response);
};

export const deleteFoodEntry = async (entryId: string): Promise<void> => {
  await apiFetch<void>(`/v1/food-entries/${entryId}`, { method: "DELETE" });
};

File: backend/app/api/routers/food_entries.py
from __future__ import annotations

import uuid
from datetime import date

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_device_id
from app.db.session import get_session
from app.schemas.food_entry import FoodEntryCreateRequest, FoodEntryResponse, FoodEntryUpdateRequest
from app.services.food_entries import (
    create_food_entry,
    get_food_entry,
    list_food_entries,
    soft_delete_food_entry,
    update_food_entry,
)


router = APIRouter(prefix="/food-entries", tags=["food-entries"])


@router.get("", response_model=list[FoodEntryResponse])
async def food_entries_list(
    day: date | None = Query(default=None),
    from_day: date | None = Query(default=None, alias="from"),
    to_day: date | None = Query(default=None, alias="to"),
    device_id: uuid.UUID = Depends(get_current_device_id),
    session: AsyncSession = Depends(get_session),
) -> list[FoodEntryResponse]:
    return await list_food_entries(session, device_id=device_id, day=day, from_day=from_day, to_day=to_day)


@router.post("", response_model=FoodEntryResponse, status_code=status.HTTP_201_CREATED)
async def food_entries_create(
    body: FoodEntryCreateRequest,
    device_id: uuid.UUID = Depends(get_current_device_id),
    session: AsyncSession = Depends(get_session),
) -> FoodEntryResponse:
    entry = await create_food_entry(
        session,
        device_id=device_id,
        product_id=body.product_id,
        portion_id=body.portion_id,
        day=body.day,
        meal_type=body.meal_type,
        amount=body.amount,
        unit=body.unit,
    )
    if entry is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Not found")
    return entry


@router.get("/{entry_id}", response_model=FoodEntryResponse)
async def food_entries_get(
    entry_id: uuid.UUID,
    device_id: uuid.UUID = Depends(get_current_device_id),
    session: AsyncSession = Depends(get_session),
) -> FoodEntryResponse:
    entry = await get_food_entry(session, device_id=device_id, entry_id=entry_id)
    if entry is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Not found")
    return entry


@router.patch("/{entry_id}", response_model=FoodEntryResponse)
async def food_entries_update(
    entry_id: uuid.UUID,
    body: FoodEntryUpdateRequest,
    device_id: uuid.UUID = Depends(get_current_device_id),
    session: AsyncSession = Depends(get_session),
) -> FoodEntryResponse:
    patch: dict = {}
    for key in ["portion_id", "meal_type", "amount", "unit"]:
        val = getattr(body, key)
        if val is not None:
            patch[key] = val

    entry = await update_food_entry(session, device_id=device_id, entry_id=entry_id, patch=patch)
    if entry is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Not found")
    return entry


@router.delete("/{entry_id}", status_code=status.HTTP_204_NO_CONTENT, response_model=None)
async def food_entries_delete(
    entry_id: uuid.UUID,
    device_id: uuid.UUID = Depends(get_current_device_id),
    session: AsyncSession = Depends(get_session),
) -> None:
    ok = await soft_delete_food_entry(session, device_id=device_id, entry_id=entry_id)
    if not ok:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Not found")

